### DISCUSSION TITLE
Default operation for array items?

### DISCUSSION THREAD

**Discussion by pivotaljohn** [k14s/ytt#254] [Ideas] [answered]
A common pain point (especially for the newly initiated) is the need to add an annotation to every &$!@ array item named in an overlay.

This is a sharp edge for folks learning about `ytt`, let alone overlays.

With Schema, all Data Values are about to become overlays (even the first Data Value document). This means for someone who is just using a `ytt` library, they'll have to know/learn that they need to include _some_ kind of matcher+operation for every array item in a data value. This feels like an unduly steep learning curve.

In the same way that we assume `@overlay/merge` (-ish) as the default for map items...

What if the default operation of array items is `overlay/append`?

---
_(last update: reflect results of design conversations from 2021-01-25)_

**Comment by cppforlife** [comment]
alternative to consider here is replacing entire array (left side) with given array (right side).

^ this might be a less surprising behaviour and may be more familiar from other tools.

**Comment by pivotaljohn** [comment]
With no objection, here, and the driving need to make the UX around Schemas as smooth as we can, we're pulling the trigger on this: https://www.pivotaltracker.com/story/show/176489687.

**Reply by pivotaljohn** [reply]
There were objections — or at least a desire to unpack the implications of this idea. There was a design discussion circa week 4 of 2021.

TL;DR.
- an implied `@overlay/replace` on the containing array is problematic
  - created a break change.
  - when expressions are present on array items (i.e. the children of the array), that exclude those items, the overall behavior changes in undesirable ways.
- an implied `@overlay/append` on the array item avoids those problems.

**Comment by joaopapereira** [comment, marked as answer]
Created issue #269 to depict the result of this conversation
